_posts/2021-04-09-springboot-autoconfigurations.md

```diff
@@ -0,0 +1,436 @@
+    </bean>
+
+    <bean id="transactionManager" class="org.springframework.orm.jpa.JpaTransactionManager">
+        <property name="entityManagerFactory" ref="entityManagerFactory" />
+        <property name="dataSource" ref="dataSource" />
+    </bean>
+
+    <tx:annotation-driven transaction-manager="transactionManager"/>
+
+ ```
+In Spring-based application, above XML configuration we have to do manually for each of the beans contained in it and wire/associate them properly.
+
+## 4. Role of Conditional Auto-Configurations with @Conditional annotation
+
+Before we understand the concept of auto-configuration, we need to understand just one important concept i.e., Spring Framework's ``@Conditional`` annotation.
+
+### 4.1 What is @Conditional?
+
+``@Conditional`` indicates that a component is only eligible for registration when all specified conditions match. A ``condition`` is any state that can be determined programmatically before the bean definition is due to the registered.
+
+```java
+
+   @Target(value = {TYPE, METHOD }
+   @Retention(value = RUNTIME)
+   @Documented
+   public @interface Conditional
+
+```
+
+The ``@Conditional`` annotation may be used in any of the following ways:
+   * as a type-level annotation on any class directly or indirectly annotated with ``@Component``, including ``@Configuration`` classes.
+   * as a meta-annotation, for the purpose of composing custom stereotype annotations.
+   * as a method-level annotation on any ``@Bean`` method.
+
+You can find more information at this link [Conditional Beans with Spring Boot!](https://reflectoring.io/spring-boot-conditionals/)
+
+
+## 5. An in-depth understanding of Spring Boot and its AutoConfigurations
+
+### 5.1 What happens internally when you start Spring Boot Application?
+
+When you start/launch the Spring Boot application there is a main() method that launches Spring Boot's enchantment.
+
+``` java
+
+package com.bsmlabs.movieservice
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class MovieServiceApplication {
+
+   public static void main(String[] args) { // (1)
+     SpringApplication.run(MovieServiceApplication.class, args);
+   }
+
+}
+
+```
+1. when we run this main method and the Tomcat server boots up, application.properties file gets read in and we can start writing @RestControllers.
+
+Let's understand internally what happens.
+
+There are many internal things that happen when we run an application, but let's have a look at three core features
+
+* **Auto-registered @PropertySources**
+
+* **Where is Spring Boot's AutoConfigurations implemented and read-in META-INF/spring.factories**
+
+* **Enhanced Conditional Support**
+
+### **5.1.1. Auto-registered @PropertySources**
+
+In Spring framework application to read in .properties files from any location we want, with the help of the ``@PropertySource`` annotation.
```

Comment:
Using PropertySources as an example is a good idea, but you'll have to explain to the reader what a PropertySource is./